library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt

```diff
@@ -822,6 +822,9 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
         LOG.v("processTouchEvent:", "start.")
         if (mState == ANIMATING) return TOUCH_STEAL
 
+        // if we are flinging, just cancel it and process touch as normal
+        cancelFling()
```

Comment:
I don't think this is needed as as soon as we find a new gesture, that will cancel the fling through `setState()`

Reply:
I think I got a little excited about the idea of supporting this behaviour with just the addition of a single line. I spend the last hours trying to figure this out...

Imo there should be only a single code section where the "fix animation" is started, and that is when the user releases touch completely. I tried to implement this but had lots of strange behavior with the animation never triggering or `NONE` states right after the animation started and even if that worked the behavior changed when an animation was already running when touching the screen. I also had a hard time understanding how and why return values of the pinch and flingscroll listener are used in the `processTouchEvent` method and how they affect the overscroll/pan fix animation.

My final resolution is that this is a topic in itself and I will not attempt to implement it in this PR. I will rework this PR and force push anything unrelated - but I have to give my brain a little rest know :smile: I would really like this to happen though as it is very unintuitive that flings and animations can not be cancelled by user touch.
I will inform you if I am ready for a new review.

Reply:
Why is that? Note that there is also a `setState(NONE)` in `onDown()`, which was put there to do this exact job. And might be causing some of the weirdness you have been experiencing.

Without trying anything I would say that that specific `setState(NONE)` is enough to cancel animations on user touch:

- Developer animations: if you replace `mCancelAnimations == true` with `mState == ANIMATING`, this check will fail at the next cycle, so that's fine. It will stop.

- Fling animations: I don't know how it works at the moment, but the flinger could also be doing the same - check that state is still FLINGING before applying the new values at each step.

These are just a few lines. Then on top of this we could also have the external API `cancelAnimations()` which implementation is basically set state to NONE if it was FLINGING or ANIMATING.

The `processTouchEvent` part is delicate indeed because it was just a lot of trial and error to get it right. I wouldn't touch that too much, I'm not surprised that you've seen weird stuff :) see if the setState during onDown is enough

Reply:
If after doing these simple things (stop animations and flings by checking `state` in their own cycles), animations are still not stopped on touch... then I would like to understand why, it's very strange.

Reply:
> Why is that? Note that there is also a `setState(NONE)` in `onDown()`, which was put there to do this exact job. And might be causing some of the weirdness you have been experiencing.

There is no `setState()` in the `onDown()` method of the `FlingScrollListener` if you mean that. And even if it was it wouldn't work because when an animation is running **every** touch input event is just discarded (first line in `processTouchEvent`).

> Without trying anything I would say that that specific `setState(NONE)` is enough to cancel animations on user touch:
> 
> * Developer animations: if you replace `mCancelAnimations == true` with `mState == ANIMATING`, this check will fail at the next cycle, so that's fine. It will stop.
> * Fling animations: I don't know how it works at the moment, but the flinger could also be doing the same - check that state is still FLINGING before applying the new values at each step.

I kind of did this, but in a different way (see other comment)
The `flingScroller` is already cancelled in `setState()`when the state changes and also works when a new fling is applied while an old fling is still animating (at least in my testing).

> These are just a few lines. Then on top of this we could also have the external API `cancelAnimations()` which implementation is basically set state to NONE if it was FLINGING or ANIMATING.

Yes I did it in exactly this way.

> The `processTouchEvent` part is delicate indeed because it was just a lot of trial and error to get it right. I wouldn't touch that too much, I'm not surprised that you've seen weird stuff :) see if the setState during onDown is enough

Well I'm glad I'm not the only one frustrated with this method then :smile:
Sadly it won't be that easy though as mentioned above.